### ISSUE TITLE
Security patches?

### ISSUE THREAD

**Issue by joepie91** [NixOS/nix#614]
This is more a design question than an actual bug, but I'm not sure where else to file it.

All dependencies in Nix are immutable by default, which is desirable in almost all cases. However, there's one case where it isn't: for security patches. It's common for a vendor to release a 'security patch'; that is, a new version of a library or piece of software, that has an identical API to the previous version, but that has an internal fix to remedy a security problem.

It's generally desirable to upgrade these dependencies without requiring anything from other vendors who depend on it - patching a library once means that every application using that library is now patched, even if the application maintainers are not paying attention or are MIA. Such a patch-all strategy is practically required to keep systems secure.

How would this be accomplished with Nix? As I understand it, you cannot forcibly change the dependencies for a package, even if the API remains 100% unchanged, because dependencies are immutable and only dependents can specify upgrades. In for example the Node.js ecosystem, this is solved using semantic versioning, but Nix does not seem to offer any such equivalent.

**Comment by joachifm** [comment]
I believe this is a non-issue. Patching the source changes the output, thus triggering a rebuild of all reverse dependencies. Some nix tools do expect the name attribute to contain version information, but version numbers have no special meaning within Nix itself.

**Comment by joepie91** [comment]
Hmm, can you elaborate on the rebuild a little? So far I've read the 'about' page and the 2004 paper, and I haven't seen any mention of triggering rebuilds. How would this work from a technical perspective, and what would a typical security patch deployment look like from the POVs of a maintainer and an end user?

**Comment by joachifm** [comment]
It helps to think about a build recipe as a pure function `F(X)`, where `X` is a set of build inputs, whose output can be used as an input to other build recipes. Due to purity, `F` only does work the first time you call it for any given `X`, otherwise it returns a cached result. Patching source code is akin to applying `F` to a different set of inputs, thus causing it do work to produce a result. This result is likely to differ from the previous result, and so any build recipe which uses the output of `F` as an input now also has to do work to produce a result, and so on (that is, patching the source has  "triggered" a chain of rebuilds).

To wit, security patches are not special and do not require special handling. You should, however, ensure that you are subscribed to an active channel and use `nix-env --leq -u '*'` to update your user profile (unsure why this is not default). I suppose the only thing which is lacking is a built-in mechanism for nagging channel subscribers about upgrading.

**Comment by joepie91** [comment]
Right. I guess I'm not so much talking about patching source, as I'm talking about patched *versions*. I'll give a (fictitious) example:

* You have `firefox-40` installed, which depends on `openssl-1.2.3`.
* A severe security issue is found in `openssl-1.2.3`, and the `openssl` maintainers publish a patched version `openssl-1.2.4` - it only patches the issue, but is otherwise fully API-compatible with `openssl-1.2.3`.
* The `firefox` maintainers are away on holiday, and do not update their package to point at the new `openssl-1.2.4`. It remains dependent on the broken `openssl-1.2.3`.

Now this is based on my understanding of Nix thus far, so it may not be 100% accurate, but deriving from the above, there would be two possibilities:

1. Users are now stuck with a known-vulnerable `openssl`, because the `firefox` package wasn't updated to point at `openssl-1.2.4` - even though it's available as a secure drop-in update. Because dependencies are immutable, users will remain vulnerable for weeks.
2. Nix has some kind of mechanism to force API-compatible dependency updates (eg. for security reasons), and updates the `firefox-40` package to point at `openssl-1.2.4` instead. The user is now secure.

In the case of scenario 2, it is critical that the user is *not* expected to have any knowledge of Nix, or even awareness that it exists. While manual patching can be expected from system administrators, that is not the case for an arbitrary home user.

Which of these two scenarios (or maybe a third I didn't think of?) is currently the case for Nix? For context, NPM (the package manager used for Node.js) solves this problem through semantic versioning queries.

**Comment by joachifm** [comment]
In nixpkgs, you typically only ever have a single top-level name for each package. Thus, `openssl` would be associated with the default (usually latest) version of OpenSSL, and all dependents would refer to this name. Changing the value associated with that name, then, would trigger a rebuild as I described above.

Be aware that "name" in this case refers to the name of the attribute in the package set, not the name attribute of the package. The name attribute of the package does not enter into this at all.

The scenario you describe would only occur if there was a top-level name associated with a specific version of OpenSSL, which was then used as an input to firefox instead of the default.

**Comment by joepie91** [comment]
I see.

1. So would a 'system-wide' (environment-wide?) update then result in `openssl` being upgraded, even if it's only ever used as a dependency and never explicitly installed by the user?
2. And would it be possible to still require an exact version of `openssl` in a hypothetical `curl` package, while `firefox` just uses whatever the latest version is?

The problem I'm seeing otherwise is that sometimes updates will (and should) involve API breaks, and that if you can't require specific versions or ranges, you might end up building against an incompatible library. That would place a library maintainer in a difficult position: either allow for automatic security updates in the future, or allow for API breaks in the future.

**Comment by joachifm** [comment]
Yes to both. It's only a question of having names for all the versions you want and then passing the correct names. You can, e.g., have one name for each major release series of a package. Of course, the maintenance burden increases drastically for each additional name you introduce, which is why you'll usually only see one per package.

Note that there is no limit to the number of versions/variants of a library that can be installed at any time. Every package refers explicitly to the specific version/variant that it was built against.

**Comment by joepie91** [comment]
I see. So for example a package following semantic versioning would have the following packages:

* semver-1.0
* semver-1.1
* semver-2.0

Does that look about right? And would it be possible to do the following as well, by 'inheriting' packages?

* semver-1.0.0
* semver-1.0.1
* semver-1.0 (always 'inherits' from the latest 1.0.x, and is updated when a new 1.0.x comes out)
* semver-1.1.0
* semver-1.1.1
* semver-1.1 (again, points at latest 1.1.x)
* semver-2.0.0
* semver-2.0 (points at latest 2.0.x)
* semver-1 (points at latest 1.x)
* semver-2 (points at latest 2.x)

And finally the third question, would it be possible to 'emulate' semantic versioning somehow, by enforcing a rule like "the latest `semver-1.0.x`, but it must be __at least__ 1.0.1" (ie. the equivalent of `~1.0.1`)?

**Comment by joachifm** [comment]
Yes to everything, except there's no meta language for describing constraints. That is, you have to
implement the constraints directly.
```nix
rec {
  # Specific versions of the foo package
  foo-2_0_1 = { ... } { };
  foo-2_0_2 = { ... } { };

  # Latest in the 2.x series (an example of "inheritance")
  foo-2 = foo-2_0_2;

  # The default version
  foo = foo-2;

  # Package which uses the latest foo
  quux = { foo = foo; } { };

  # Package which uses the latest foo 2.x
  quuz = { foo = foo-2; } { };
}
```
and so on. Now, adding a name for `foo` version `2.0.3` would require you to also update the binding for `foo-2` to maintain the relation. If you added a `foo-3` and updated the default to refer to it, then `quux` would automatically use the new `foo-3`, whereas `quuz` would stick with `foo-2`. To be clear, the attribute names can be whatever you want.

If you really want to, you could encode constraints as functions taking a package set as an input and returning the package satisfying the constraint (as in `head (filter myPredicate (attrSetAsList pkgs))`), but I fail to see the point/value in doing that.

**Comment by vcunat** [comment]
I didn't read this discussion, but you should know about https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/replace-dependency.nix. Once there were very long discussions around this topic, probably on the mailing-list...
